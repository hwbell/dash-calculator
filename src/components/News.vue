<template>
  <div id="news-component">
    <div v-if="articles != null" class="news-articles-holder left-all-col">
      <div style="width: 100%">
        <div class="left-all-row" style="width: 100%">
          <p style="margin: 0px 10px 0 0">
            <strong
              >Top News since
              <strong style="color: #007bff">{{ sorter }}</strong>
            </strong>
          </p>
          <button
            @click="sortArticles(lastDay)"
            class="btn btn-sm btn-outline-primary"
          >
            last day
          </button>
          <button
            @click="sortArticles(lastWeek)"
            class="btn btn-sm btn-outline-primary"
          >
            last week
          </button>
          <button
            @click="sortArticles(lastMonth)"
            class="btn btn-sm btn-outline-primary"
          >
            last month
          </button>
        </div>

        <div class="left-all-row">
          <input
            class="search-input"
            @keyup.enter="fireSearch()"
            v-model="searchQuery"
            type="text"
          />
          <button @click="fireSearch()" class="btn btn-sm btn-primary">
            search
          </button>
          <div v-if="loading" class="spinner-border text-primary" role="status">
          </div>
        </div>
      </div>

      <div
        v-for="(article, ind) in articles"
        v-bind:key="ind"
        class="news-article shadow-card"
      >
        <div class="space-all-row">
          <div class="left-all-col">
            <p class="article-title">{{ article.title }}</p>
            <p>{{ article.publishedAt.slice(0, 10) }}</p>
          </div>
          <img
            class="article-image"
            v-bind:src="article.urlToImage"
            alt="article image"
          />
        </div>
        <p>
          <span class="article-author" v-if="article.author">{{ article.author }}, </span
          ><span v-if="article.source.name">
            from {{ article.source.name }}</span
          >
        </p>
        <p v-if="article.content" class="article-description">
          <span>{{ article.content.slice(0, -(article.content.length-article.content.indexOf('['))) }}</span>
        </p>
        <a v-bind:href="article.url" target="_blank" alt="article link"
          >&nbsp;&nbsp;&nbsp;read full article</a
        >
      </div>
    </div>
  </div>
</template>

<script>
// import { fetchArticles } from "../tools/newsapi";

export default {
  name: "News",
  // props: {
  // msg: String,
  // },
  // components: {ExpressionWindow},
  data: function() {
    const moment = require("moment");
    let today = moment().format("YYYY-MM-DD");
    let lastDay = moment()
      .subtract(1, "d")
      .format("YYYY-MM-DD");

    let lastWeek = moment()
      .subtract(7, "d")
      .format("YYYY-MM-DD");
    let lastMonth = moment()
      .subtract(30, "d")
      .format("YYYY-MM-DD");

    return {
      loading: true,
      searchQuery: "covid-19",
      sorter: lastWeek,
      message: "This is a news component for us to start with!",
      articles: [],
      today,
      lastDay,
      lastWeek,
      lastMonth,
    };
  },
  created: function() {
    this.getArticles(this.searchQuery, this.sorter);
  },
  computed: {
    // a computed getter
  },
  methods: {
    getArticles: async function(query, fromDate) {
      if (Boolean(fromDate) === false) {
        fromDate = this.lastDay;
      }
      let serverResponse;

      // construct the url with the query and to / from dates
      const searchUrl = `https://cors-anywhere.herokuapp.com/https://newsapi.org/v2/everything?q=${query}&from=${fromDate}&to=${this.today}`;

      // check localstorage first
      if (Boolean(localStorage.getItem(searchUrl)) === true) {
        console.log(`found news search for ${searchUrl} in local storage`);
        serverResponse = JSON.parse(localStorage.getItem(searchUrl));
        this.articles = serverResponse.articles;
        this.loading = false;

        return serverResponse;
      }

      let headers = {
        Authorization: process.env.VUE_APP_NEWSAPIKEY,
        'Access-Control-Allow-Origin': '*'
      };
      this.loading = true;

      serverResponse = await fetch(searchUrl, {
        method: "GET",
        headers,
      })
        .then((response) => response.json())
        .then((response) => {
          return response;
        })
        .catch((err) => console.log(err));

      console.log(serverResponse);

      // save to state
      this.loading = false;
      this.articles = serverResponse.articles;
      // save to local storage
      localStorage.setItem(searchUrl, JSON.stringify(serverResponse));
      return serverResponse;
    },
    sortArticles: async function(sorter) {
      this.loading = true;
      this.sorter = sorter;
      console.log(`re-fetching articles since ${sorter}`);
      this.getArticles(this.searchQuery, sorter);
    },
    fireSearch: function() {
      this.loading = true;
      console.log(`searching for ${this.searchQuery}`);
      this.getArticles(this.searchQuery, this.sorter);
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
h3 {
  margin: 40px;
}
.btn {
  margin: 4px;
}
a {
  font-weight: bolder;
}
.spinner-border {
  font-size: 12pt;
  margin: 0 10px;
}
#news-component {
  /* width: 100%; */
  padding: 20px;
  background-color: white;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: flex-start;
}
.search-input {
  /* width: 100%; */
  border-radius: 4px;
  margin: 10px 0;
  border: 1px solid rgb(170, 170, 170);
}
.news-article {
  background-color: white;
  width: 100%;
  margin: 20px 0;
  padding: 20px;
  border-radius: 2px;
  /* border: 1px solid rgb(207, 206, 206); */
}
.article-image {
  /* width: 100px; */
  height: 100px;
  border-radius: 10px;
}
.article-title {
  color: black;
  font-weight: bolder;
  font-size: 14pt;
}
.article-author {
  color: rgb(63, 63, 63);
  font-weight: bolder;
  font-size: 12pt;
}
.article-description {
  color: black;
  font-size: 12pt;
}
</style>
