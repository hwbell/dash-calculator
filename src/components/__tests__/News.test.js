import { mount } from '@vue/test-utils'
import News from '../News.vue'
import Vue from "vue";

const message = 'This is a base component for us to start with!';

const mockArticles = [
  {
    source: { id: null, name: 'TheChronicleHerald.ca' },
    author: 'Joey Roulette',
    title: "Musk's SpaceX set for debut astronaut mission, renewing NASA's crewed launch program - TheChronicleHerald.ca",
    description: "<ol><li>Musk's SpaceX set for debut astronaut mission, renewing NASA's crewed launch program  TheChronicleHerald.ca\r\n" +
      '</li><li>NASA astronauts go back to the future with capsule launch  CTV News\r\n' +
      '</li><li>SpaceX launch with NASA astronauts will feature retro-s…',
    url: 'https://www.thechronicleherald.ca/news/world/musks-spacex-set-for-debut-astronaut-mission-renewing-nasas-crewed-launch-program-453854/',
    urlToImage: 'https://www.thechronicleherald.ca/media/photologue/photos/cache/musks-spacex-set-for-debut-astronaut-mission-renewing-nasas-crewed-launch-program_medium',
    publishedAt: '2020-05-26T11:14:39Z',
    content: 'By Joey Roulette\r\n' +
      "CAPE CANAVERAL, Fla. (Reuters) - Entrepreneur Elon Musk's SpaceX is set to launch two American astronauts to the International Space Station on Wednesday from NASA's Kennedy Space C… [+1974 chars]"
  },
  {
    source: { id: null, name: 'Global News' },
    author: 'Curt Anderson And Brian Mahoney',
    title: 'U.S. Memorial Day crowds small, subdued amid coronavirus pandemic - Global News',
    description: '<ol><li>U.S. Memorial Day crowds small, subdued amid coronavirus pandemic  Global News\r\n' +
      '</li><li>US marks quiet Memorial Day as coronavirus deaths near 100,000  Al Jazeera English\r\n' +
      '</li><li>What Is the Difference Between Memorial Day and Veterans Day?  Newswe…',
    url: 'http://globalnews.ca/news/6981600/coronavirus-memorial-day-warnings/',
    urlToImage: 'https://shawglobalnews.files.wordpress.com/2020/05/1000-32.jpeg?quality=70&strip=all&w=720&h=379&crop=1',
    publishedAt: '2020-05-25T11:01:48Z',
    content: 'Americans settled for small processions and online tributes instead of parades Monday as they observed Memorial Day in the shadow of the pandemic, which forced communities to honour the nation’s mili… [+5855 chars]'
  }
]
const serverResponse = {
  articles: mockArticles
}
describe('News', () => {

  let wrapper;
  beforeAll( async () => {
    wrapper = mount(News, {
      data: function() {
        return {
          test: true,
          message: "This is a news component for us to start with!",
        };
      },
      methods: {
        getArticles: function () {
          Promise.resolve(serverResponse)
            .then((response) => {
              wrapper.vm.articles = response.articles;
            })
        }
      }
    })
  })

  test('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
  
  test('renders the correct content', () => {
    let newsDiv = wrapper.findAll('.news-component');
    expect(newsDiv.length).toBe(1);
    // expect(messageDiv.text()).toBe('This is a base component for us to start with!')
  })

  test(`renders an article for each result`, async () => {
    expect(wrapper.vm.articles.length).toBe(2)

    expect(wrapper.findAll('.news-article').length).toBe(2)
  })  

})