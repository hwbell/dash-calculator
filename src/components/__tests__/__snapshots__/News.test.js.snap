// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`News renders correctly 1`] = `
<div
  id="news-component"
>
  <div
    class="news-articles-holder left-all-col"
  >
    <div
      style="width: 100%;"
    >
      <div
        class="left-all-row"
        style="width: 100%;"
      >
        <p
          style="margin: 0px 10px 0px 0px;"
        >
          <strong>
            Top News since
            
            <strong
              style="color: rgb(0, 123, 255);"
            >
              2020-05-21
            </strong>
          </strong>
        </p>
         
        <button
          class="btn btn-sm btn-outline-primary"
        >
          
          last day
        
        </button>
         
        <button
          class="btn btn-sm btn-outline-primary"
        >
          
          last week
        
        </button>
         
        <button
          class="btn btn-sm btn-outline-primary"
        >
          
          last month
        
        </button>
      </div>
       
      <div
        class="left-all-row"
      >
        <input
          class="search-input"
          type="text"
        />
         
        <button
          class="btn btn-sm btn-primary"
        >
          
          search
        
        </button>
         
        <div
          class="spinner-border text-primary"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading...
          </span>
        </div>
      </div>
    </div>
     
    <div
      class="news-article shadow-card"
    >
      <div
        class="space-all-row"
      >
        <div
          class="left-all-col"
        >
          <p
            class="article-title"
          >
            Musk's SpaceX set for debut astronaut mission, renewing NASA's crewed launch program - TheChronicleHerald.ca
          </p>
           
          <p>
            2020-05-26
          </p>
        </div>
         
        <img
          alt="article image"
          class="article-image"
          src="https://www.thechronicleherald.ca/media/photologue/photos/cache/musks-spacex-set-for-debut-astronaut-mission-renewing-nasas-crewed-launch-program_medium"
        />
      </div>
       
      <p
        class="article-author"
      >
        <span>
          Joey Roulette, 
        </span>
        <span>
          
          from TheChronicleHerald.ca
        </span>
      </p>
       
      <p
        class="article-description"
      >
        <span>
          By Joey Roulette
CAPE CANAVERAL, Fla. (Reuters) - Entrepreneur Elon Musk's SpaceX is set to launch two American astronauts to the International Space Station on Wednesday from NASA's Kennedy Space C… 
        </span>
      </p>
       
      <a
        alt="article link"
        href="https://www.thechronicleherald.ca/news/world/musks-spacex-set-for-debut-astronaut-mission-renewing-nasas-crewed-launch-program-453854/"
        target="_blank"
      >
           read full article
      </a>
    </div>
    <div
      class="news-article shadow-card"
    >
      <div
        class="space-all-row"
      >
        <div
          class="left-all-col"
        >
          <p
            class="article-title"
          >
            U.S. Memorial Day crowds small, subdued amid coronavirus pandemic - Global News
          </p>
           
          <p>
            2020-05-25
          </p>
        </div>
         
        <img
          alt="article image"
          class="article-image"
          src="https://shawglobalnews.files.wordpress.com/2020/05/1000-32.jpeg?quality=70&strip=all&w=720&h=379&crop=1"
        />
      </div>
       
      <p
        class="article-author"
      >
        <span>
          Curt Anderson And Brian Mahoney, 
        </span>
        <span>
          
          from Global News
        </span>
      </p>
       
      <p
        class="article-description"
      >
        <span>
          Americans settled for small processions and online tributes instead of parades Monday as they observed Memorial Day in the shadow of the pandemic, which forced communities to honour the nation’s mili… 
        </span>
      </p>
       
      <a
        alt="article link"
        href="http://globalnews.ca/news/6981600/coronavirus-memorial-day-warnings/"
        target="_blank"
      >
           read full article
      </a>
    </div>
  </div>
</div>
`;
